sw/airborne/arch/chibios/modules/actuators/esc_dshot.c

```diff
@@ -0,0 +1,482 @@
+{
+  return &driver->dshotMotors.dt[index];
+}
+
+
+/*
+#                 _ __           _                    _
+#                | '_ \         (_)                  | |
+#                | |_) |  _ __   _   __   __   __ _  | |_     ___
+#                | .__/  | '__| | |  \ \ / /  / _` | | __|   / _ \
+#                | |     | |    | |   \ V /  | (_| | \ |_   |  __/
+#                |_|     |_|    |_|    \_/    \__,_|  \__|   \___|
+*/
+static DshotPacket makeDshotPacket(const uint16_t _throttle, const bool tlmRequest)
+{
+  DshotPacket dp = {.throttle = _throttle,
+                    .telemetryRequest = (tlmRequest ? 1 : 0),
+                    .crc = 0
+                   };
+
+  // compute checksum
+  uint16_t csum = (_throttle << 1) | dp.telemetryRequest;
+  for (int i = 0; i < 3; i++) {
+    dp.crc ^=  csum;   // xor data by nibbles
+    csum >>= 4;
+  }
+
+  return dp;
+}
+
+static inline void setDshotPacketThrottle(DshotPacket *const dp, const uint16_t throttle)
+{
+  dp->throttle = throttle;
+  dp->telemetryRequest = 0;
+}
+
+static inline void setDshotPacketTlm(DshotPacket *const dp, const bool tlmRequest)
+{
+  dp->telemetryRequest =  tlmRequest ? 1 : 0;
+}
+
+static void buildDshotDmaBuffer(DshotPackets *const dsp,  DshotDmaBuffer *const dma, const size_t timerWidth)
+{
+  for (size_t chanIdx = 0; chanIdx < DSHOT_CHANNELS; chanIdx++) {
+    // compute checksum
+    DshotPacket *const dp = &dsp->dp[chanIdx];
+    dp->crc = 0;
+    uint16_t csum = (dp->throttle << 1) | dp->telemetryRequest;
+    for (int i = 0; i < 3; i++) {
+      dp->crc ^=  csum;   // xor data by nibbles
+      csum >>= 4;
+    }
+    // generate pwm frame
+    for (size_t bitIdx = 0; bitIdx < DSHOT_BIT_WIDTHS; bitIdx++) {
+      const uint16_t value = dp->rawFrame &
+                             (1 << ((DSHOT_BIT_WIDTHS - 1) - bitIdx)) ?
+                             DSHOT_BIT1_DUTY : DSHOT_BIT0_DUTY;
+      if (timerWidth == 2) {
+        dma->widths16[bitIdx][chanIdx] = value;
+      } else {
+#if DSHOT_AT_LEAST_ONE_32B_TIMER
+        dma->widths32[bitIdx][chanIdx] = value;
+#else
+        chSysHalt("use of 32 bit timer implies to define DSHOT_AT_LEAST_ONE_32B_TIMER to TRUE");
+#endif
+      }
+    }
+    // the bits for silence sync in case of continous sending are zeroed once at init
+  }
+}
+
+static inline uint8_t updateCrc8(uint8_t crc, uint8_t crc_seed)
+{
+  uint8_t crc_u = crc;
+  crc_u ^= crc_seed;
+
+  for (int i = 0; i < 8; i++) {
+    crc_u = (crc_u & 0x80) ? 0x7 ^ (crc_u << 1) : (crc_u << 1);
+  }
+
+  return (crc_u);
+}
+
+static uint8_t calculateCrc8(const uint8_t *Buf, const uint8_t BufLen)
+{
+  uint8_t crc = 0;
+  for (int i = 0; i < BufLen; i++) {
+    crc = updateCrc8(Buf[i], crc);
+  }
+
+  return crc;
+}
+
+__attribute__((const))
+static size_t   getTimerWidth(const PWMDriver *pwmp)
+{
+  (void) pwmp;
+
+  return (0
+#if STM32_PWM_USE_TIM2
+          || (pwmp == &PWMD2)
+#endif
+#if STM32_PWM_USE_TIM5
```

Comment:
Only these timers?

Reply:
yes, those two are the only 32 bits timers, same F4 and F7